### PR TITLE
contikimac: a ContikiMAC compatible netdev layer

### PR DESCRIPTION
**Pull request by jnohlgard**
### Contribution description

Introduce a ContikiMAC compatible radio duty cycling MAC layer as a netdev layer.
The implementation uses netdev layers to wrap an underlying netdev transceiver, which means that from the gnrc_netif module the ContikiMAC layer looks and behaves as if it were a normal netdev transceiver driver.
The primary goal is to reduce power consumption, and as a proof of concept for the idea of implementing MAC layer as a netdev layer. 
The module documentation is quite extensive, see contikimac.h for details. 

### Limitations

The implementation has a few minor direct dependencies/interactions against gnrc_netif, but this situation can be improved as the netdev layer implementation is evolved (#7736, #9417 etc) 
Not yet verified against other implementations due to lack of time. Only tested against itself between samr21-xpro and frdm-kw41z.
The implementation uses the event queue introduced in #9326 to handle events from timers and handle feedback from the underlying device. The send method hijacks the event loop internally to avoid races between the IPC messages and the frame transmit algorithm. 

TODO add some more text from the module documentation. 

### Comparison to lwmac, gomach implementations

This is implemented as a netdev layer and therefore is less intrusive and does not require modifications to the network stack (gnrc). The only requirement is a netdev compatible transceiver and a working xtimer configuration. For better power savings, the platform should be configured to use a low power timer with xtimer. This is the default for frdm-kw41z, but most other boards need a configuration modification. This should only affect CPU power consumption, it should not affect communications in any way. 

### Issues/PRs references

Depends on ~~#9469~~, ~~#9470~~, ~~#9471~~, #9326  
Also useful #7577 
Description of the algorithm: http://dunkels.com/adam/dunkels11contikimac.pdf

### CODE REVIEW

**Comment by miri64** [comment]
I will have a look at it, but it might take some time.

**Comment by gebart** [comment]
@miri64 no rush, I realize that there is so much work happening in the netdev and netif_ieee802154 modules so this may need to be refactored a few times before it can be merged.

File: sys/net/link_layer/contikimac/contikimac.c [hunk 1]
```diff
@@ -0,0 +1,1139 @@
...
+    res = lower->driver->set(lower, NETOPT_STATE, &state_listen, sizeof(state_listen));
+    if (res < 0) {
+        DEBUG("contikimac(%d): Failed setting NETOPT_STATE_IDLE: %d\n",
+                  thread_getpid(), res);
+        return;
+    }
+    /* Set timeout in case we only detected noise */
+    TRACE("L");
+    netopt_state_t state;
+    res = lower->driver->get(lower, NETOPT_STATE, &state, sizeof(state));
```
[the remark this answers is not in the document]
Comment (miri64) [reply sys/net/link_layer/contikimac/contikimac.c]: Also, state is only used in the DEBUG statement below. So is this function call really necessary?

[the remark this answers is not in the document]
Comment (gebart) [reply sys/net/link_layer/contikimac/contikimac.c]: This whole state printout is a leftover from a particularly frustrating debugging session. We can remove the whole piece.

**Comment by gebart** [comment]
rebased on the latest version of #9326 
Addressed @codacy-bot comment

**Comment by miri64** [comment]
@jia200x is this on the agenda for a `netif` / `ieee8021154_submac` rework?

**Comment by jia200x** [comment]
> @jia200x is this on the agenda for a netif / ieee8021154_submac rework?

from what I see here `netdev` layer part is not needed anymore since the Radio HAL already supports Duty Cycle MAC layers on top. We would need to rewrite the netdev adoption part and I guess it should be enough for integrating this feature

**Pull request closed without merging**